# Задание №3 для тестового задания Mailfit

In [4]:
import pandas as pd
import numpy as np

from datetime import datetime

import numpy as np
import matplotlib.pyplot as plt
from matplotlib import transforms
import seaborn as sns
import plotly.express as px
from plotly import graph_objects as go 

In [28]:
data = pd.read_excel("C:/Users/azaxa/Desktop/Data для тестового.xlsx", sheet_name="Data")
data

Название рассылки      Название кампании Направление    Месяц  \
0      Название рассылки 1    Название кампании 1       Email  Октябрь   
1     Название рассылки 10   Название кампании 10       Email   Ноябрь   
2    Название рассылки 100  Название кампании 100       Email   Апрель   
3    Название рассылки 101  Название кампании 101       Email   Апрель   
4    Название рассылки 102  Название кампании 102       Email   Апрель   
..                     ...                    ...         ...      ...   
213   Название рассылки 95   Название кампании 95       Email     Март   
214   Название рассылки 96   Название кампании 96       Email     Март   
215   Название рассылки 97   Название кампании 97       Email     Март   
216   Название рассылки 98   Название кампании 98       Email     Март   
217   Название рассылки 99   Название кампании 99       Email     Март   

          Дата   Год  Номер недели  День недели   День недели.1     Время  \
0   2021-10-27  2021            43            3        03-среда  19:24:00   
1   2021-11-05  2021            45            5      05-пятница  12:02:00   
2   2022-04-11  2022            15            1  01-понедельник  16:26:00   
3   2022-04-12  2022            15            2      02-вторник  16:26:00   
4   2022-04-13  2022            15            3        03-среда  16:26:00   
..         ...   ...           ...          ...             ...       ...   
213 2022-03-24  2022            12            4      04-четверг  16:01:00   
214 2022-03-24  2022            12            4      04-четверг  16:01:00   
215 2022-03-24  2022            12            4      04-четверг  15:04:00   
216 2022-03-28  2022            13            1  01-понедельник  16:01:00   
217 2022-03-28  2022            13            1  01-понедельник  16:02:00   

     ...  Доставлено       Открытия         Клики  Баунсы (Все ошибки)  \
0    ...   741749.55  148349.910000  17801.989200             39039.45   
1    ...   683401.50  123012.270000  11071.104300             35968.50   
2    ...  1141344.25  182615.080000  15339.666720             60070.75   
3    ...  1324135.65  264827.130000  10328.258070             69691.35   
4    ...  1212979.95  218336.391000  15720.220152             63841.05   
..   ...         ...            ...           ...                  ...   
213  ...  2421707.21  422103.566703  16462.039101             36878.79   
214  ...  1077267.94  119038.107370   8570.743731             21985.06   
215  ...  2012963.12  185192.607040  22223.112845             41080.88   
216  ...   992259.45  109644.669225   9868.020230             15110.55   
217  ...  2425911.60  436664.088000  36679.783392             49508.40   

         Отписки  UTM Метка  Пользователей на сайте  Воронка продаж. Шаг 1  \
0     7417.49550    Метка 1            16377.830064            6336.582452   
1     6834.01500   Метка 10            10296.126999            3558.341491   
2    11413.44250  Метка 100            13959.096715            3369.725947   
3    13241.35650  Метка 101             8985.584521            5116.391826   
4    12129.79950  Метка 102            10846.951905            2816.953410   
..           ...        ...                     ...                    ...   
213  21795.36489   Метка 95            14321.974018            8154.932006   
214   9695.41146   Метка 96             5913.813174            1535.817281   
215  18116.66808   Метка 97            21556.419459            8991.182557   
216   8930.33505   Метка 98             8387.817196            5082.178439   
217  21833.20440   Метка 99            33745.400721           13056.095539   

     Воронка продаж. Шаг 2 Воронка продаж. Шаг 3  
0              6209.850803           5154.176166  
1              3095.757097           2538.520820  
2              2864.267055           2262.770973  
3              4277.303567           3207.977675  
4              2287.366169           1669.777303  
..                     ...                   ...  
213         

In [29]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 218 entries, 0 to 217
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Название рассылки       218 non-null    object        
 1   Название кампании       218 non-null    object        
 2   Направление             218 non-null    object        
 3   Месяц                   218 non-null    object        
 4   Дата                    218 non-null    datetime64[ns]
 5   Год                     218 non-null    int64         
 6   Номер недели            218 non-null    int64         
 7   День недели             218 non-null    int64         
 8   День недели.1           218 non-null    object        
 9   Время                   218 non-null    object        
 10  Веб-версия              218 non-null    object        
 11  Тема письма             218 non-null    object        
 12  Сегмент                 218 non-null    object    

In [55]:
data['Delivery rate'] = (data['Доставлено'] / data['Отправлено']) * 100
data['Open rate'] = (data['Открытия'] / data['Доставлено']) * 100
data['Click to Open rate'] = (data['Клики'] / data['Открытия']) * 100
data['Unsubscribe rate'] = (data['Отписки'] / data['Доставлено']) * 100

In [67]:
delivery_rate_data = data[['Тема письма','Delivery rate']].sort_values('Delivery rate',ascending=False)
delivery_rate_data

Тема письма  Delivery rate
96   Тема письма 186           98.5
51   Тема письма 145           98.5
132  Тема письма 218           98.5
186   Тема письма 70           98.5
123   Тема письма 21           98.5
..               ...            ...
24   Тема письма 120           95.0
25   Тема письма 121           95.0
26   Тема письма 122           95.0
1     Тема письма 10           95.0
0      Тема письма 1           95.0

[218 rows x 2 columns]

In [103]:
delivery_top = delivery_rate_data[delivery_rate_data['Delivery rate'] == 98.5]

Тема письма  Delivery rate
42   Тема письма 137           98.5
168   Тема письма 54           98.5
39   Тема письма 134           98.5
180   Тема письма 65           98.5
141    Тема письма 3           98.5
87   Тема письма 178           98.5
90   Тема письма 180           98.5
84   Тема письма 175           98.5
48   Тема письма 142           98.5
45    Тема письма 14           98.5
144   Тема письма 32           98.5
147   Тема письма 35           98.5
177   Тема письма 62           98.5
150   Тема письма 38           98.5
174    Тема письма 6           98.5
57   Тема письма 150           98.5
171   Тема письма 57           98.5
156   Тема письма 43           98.5
120  Тема письма 207           98.5
165   Тема письма 51           98.5
33   Тема письма 129           98.5
138   Тема письма 27           98.5
189   Тема письма 73           98.5
102  Тема письма 191           98.5
216   Тема письма 98           98.5
117  Тема письма 204           98.5
213   Тема письма 95           98.5
114  Тема письма 201           98.5
210   Тема письма 92           98.5
126  Тема письма 212           98.5
207    Тема письма 9           98.5
108  Тема письма 197           98.5
204   Тема письма 87           98.5
30   Тема письма 126           98.5
201   Тема письма 84           98.5
105  Тема письма 194           98.5
159   Тема письма 46           98.5
192   Тема письма 76           98.5
198   Тема письма 81           98.5
195   Тема письма 79           98.5
93   Тема письма 183           98.5
135   Тема письма 24           98.5
27   Тема письма 123           98.5

In [66]:
open_rate_data = data[['Тема письма','Open rate']].sort_values('Open rate',ascending=False)
open_rate_data

Тема письма  Open rate
0      Тема письма 1       20.0
3    Тема письма 101       20.0
135   Тема письма 24       18.0
55   Тема письма 149       18.0
65   Тема письма 158       18.0
..               ...        ...
161   Тема письма 48        9.2
115  Тема письма 202        9.2
112   Тема письма 20        9.2
140   Тема письма 29        9.2
148   Тема письма 36        9.2

[218 rows x 2 columns]

In [65]:
click_to_open_rate_data = data[['Тема письма','Click to Open rate']].sort_values('Click to Open rate',ascending=False)
click_to_open_rate_data

Тема письма  Click to Open rate
120  Тема письма 207                12.0
5    Тема письма 103                12.0
145   Тема письма 33                12.0
25   Тема письма 121                12.0
160   Тема письма 47                12.0
..               ...                 ...
188   Тема письма 72                 3.9
68   Тема письма 160                 3.9
18   Тема письма 115                 3.9
118  Тема письма 205                 3.9
88   Тема письма 179                 3.9

[218 rows x 2 columns]

In [69]:
unsubscribe_rate_data = data[['Тема письма','Unsubscribe rate']].sort_values('Unsubscribe rate',ascending=True)
unsubscribe_rate_data

Тема письма  Unsubscribe rate
126  Тема письма 212              0.15
134   Тема письма 23              0.15
132  Тема письма 218              0.15
131  Тема письма 217              0.15
130  Тема письма 216              0.15
..               ...               ...
145   Тема письма 33             13.00
146   Тема письма 34             13.00
147   Тема письма 35             13.00
138   Тема письма 27             13.00
148   Тема письма 36             13.00

[218 rows x 2 columns]

In [120]:
data_metric = data[['Тема письма',
                    'Delivery rate',
                    'Open rate',
                    'Click to Open rate',
                    'Unsubscribe rate']].sort_values(by = ['Delivery rate',
                                                           'Open rate',
                                                           'Click to Open rate',
                                                            'Unsubscribe rate'], ascending = [False,False,False,True])
data_metric.head(5)

Тема письма  Delivery rate  Open rate  Click to Open rate  \
153   Тема письма 40           98.5      18.00                 3.9   
123   Тема письма 21           98.5      17.43                 3.9   
36   Тема письма 131           98.5      15.95                 9.0   
111    Тема письма 2           98.5      11.05                 9.0   
51   Тема письма 145           98.5      11.05                 9.0   

     Unsubscribe rate  
153             11.00  
123              0.29  
36               0.60  
111              0.29  
51               0.57

In [119]:
datas = data[data['Unsubscribe rate'] < 0.5]
data_metric2 = datas[['Тема письма',
                    'Delivery rate',
                    'Open rate',
                    'Click to Open rate',
                    'Unsubscribe rate']].sort_values(by = ['Delivery rate',
                                                           'Open rate',
                                                           'Click to Open rate',
                                                            'Unsubscribe rate'], ascending = [False,False,False,True])
data_metric2.head(5)

Тема письма  Delivery rate  Open rate  Click to Open rate  \
123   Тема письма 21           98.5      17.43                 3.9   
111    Тема письма 2           98.5      11.05                 9.0   
129  Тема письма 215           98.5      11.05                 7.2   
132  Тема письма 218           98.5      11.05                 8.4   
117  Тема письма 204           98.5      18.00                 8.4   

     Unsubscribe rate  
123              0.29  
111              0.29  
129              0.15  
132              0.15  
117              0.25

### Лучшая тема по Delivery rate - в датасете delivery_top
### Лучшая тема по open_rate_data - Тема письма 1	и Тема письма 101
### Лучшая тема по click_to_open_rate_data - Тема письма 207	
### Лучшая тема по unsubscribe_rate_data - Тема письма 212


### По общим показателям лучшая тема - Тема письма 40 - хотя по Unsubscribe rate это нелучшая тема.                  Если рассматривать темы с Unsubscribe rate < 0.5 то лучшей темой будет - Тема письма 21

In [93]:
open_data = data.groupby('День недели.1').agg(cnt_open=('Открытия','sum')).round().sort_values('cnt_open')

In [95]:
open_data

cnt_open
День недели.1             
06-суббота        742225.0
02-вторник       4540914.0
01-понедельник   7261849.0
03-среда         7573699.0
04-четверг      11333324.0
05-пятница      11379792.0

In [97]:
open_rate = data.groupby('День недели.1').agg(cnt_open=('Open rate','mean')).round().sort_values('cnt_open')

In [99]:
open_rate

cnt_open
День недели.1           
02-вторник          13.0
01-понедельник      14.0
03-среда            14.0
04-четверг          14.0
05-пятница          14.0
06-суббота          14.0

In [98]:
click_to_Open = data.groupby('День недели.1').agg(cnt_open=('Click to Open rate','mean')).round().sort_values('cnt_open')

In [100]:
click_to_Open

cnt_open
День недели.1           
06-суббота           7.0
01-понедельник       8.0
03-среда             8.0
04-четверг           8.0
05-пятница           8.0
02-вторник           9.0

## По кол-ву открытий и по open_rate лучшим днем недели является пятница, но по показателю click_to_оpen лучшим днем является вторник.
### Стоит тестировать дни для отправки рассылок, так же важно проанализировать зависимость кол-ва открытий и тем писем, так как, возможно, что в конце рабочей недели люди не открывают письма, требующие высокой вовлеченности, в отличие от вторника, середины рабочей недели, когда готовность принимать решения выше. Но так же возможно и обратное, что в пятницу люди открывают чаще письма с отвлеченными темами, то есть в конце рабочей недели находят время для разбора рассылок. 